# Project ideas

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import json
import time
!pip install openpyxl
import openpyxl 

Defaulting to user installation because normal site-packages is not writeable


## Import JSON file to pandas DataFrame:

In [3]:
df = pd.read_json("kym.json")

In [4]:
df.shape

(28799, 16)

In [5]:
#df.describe # no nice output as there are no numeric values for most of the cols

In [6]:
df.dtypes

title                     object
url                       object
last_update_source         int64
category                  object
template_image_url        object
meta                      object
ld                        object
added                    float64
details                   object
content                   object
tags                      object
additional_references     object
search_keywords           object
parent                    object
siblings                  object
children                  object
dtype: object

In [7]:
df.category.unique()

array(['Meme', 'Subculture', 'Event', 'Culture', 'Site', 'Person'],
      dtype=object)

In [8]:
counts_per_categories = df.groupby('category').count()

## Filter out only memes

In [9]:
memes = df[df.category=="Meme"].reset_index(drop=True)

In [10]:
memes.shape

(15406, 16)

Drop the column category?

In [11]:
memes.head()

title  \
0  This is Relevant To My Interests   
1                        ROFLcopter   
2                Bitches Don't Know   
3               Leave Britney Alone   
4                            O RLY?   

                                                 url  last_update_source  \
0  https://knowyourmeme.com/memes/this-is-relevan...          1547002898   
1          https://knowyourmeme.com/memes/roflcopter          1591400337   
2   https://knowyourmeme.com/memes/bitches-dont-know          1607986999   
3  https://knowyourmeme.com/memes/leave-britney-a...          1613129339   
4               https://knowyourmeme.com/memes/o-rly          1615643899   

  category                                 template_image_url  \
0     Meme  https://i.kym-cdn.com/entries/icons/original/0...   
1     Meme  https://i.kym-cdn.com/entries/icons/original/0...   
2     Meme  https://i.kym-cdn.com/entries/icons/original/0...   
3     Meme  https://i.kym-cdn.com/entries/icons/original/0...   
4     Meme  https://i.kym-cdn.com/entries/icons/original/0...   

                                                meta  \
0  {'og:title': 'This is Relevant To My Interests...   
1  {'og:title': 'ROFLcopter', 'og:site_name': 'Kn...   
2  {'og:title': 'Bitches Don't Know', 'og:site_na...   
3  {'og:title': 'Leave Britney Alone', 'og:site_n...   
4  {'og:title': 'O RLY?', 'og:site_name': 'Know Y...   

                                                  ld         added  \
0  {'@context': 'http://schema.org', '@type': 'Br...  1.229113e+09   
1  {'@context': 'http://schema.org', '@type': 'Br...  1.229113e+09   
2  {'@context': 'http://schema.org', '@type': 'Br...  1.229113e+09   
3  {'@context': 'http://schema.org', '@type': 'Br...  1.229113e+09   
4  {'@context': 'http://schema.org', '@type': 'Br...  1.229113e+09   

                                             details  \
0  {'status': 'confirmed', 'origin': 'I Can Has C...   
1  {'status': 'confirmed', 'origin': 'Warcraft II...   
2  {'status': 'confirmed', 'origin': 'MySpace', '...   
3  {'status': 'confirmed', 'origin': 'YouTube', '...   
4  {'status': 'confirmed', 'origin': 'Something A...   

                                             content  \
0  {'about': {'text': ['"This is Relevant To My I...   
1  {'about': {'text': ['ROFLcopter is an Internet...   
2  {'about': {'text': ['Bitches Don't Know is an ...   
3  {'about': {'text': ['"Leave Britney Alone" is ...   
4  {'about': {'text': ['O RLY is internet slang f...   

                                                tags  \
0  [image macros, comment, comments, approval, im...   
1   [ascii, animated gif, rofl, acronym, initialism]   
2  [myspace, 4chan, parody, know, bitches, dont, ...   
3  [youtube, rant, britney spears, response, frea...   
4  [image macro, owl, orly, o rly, negative, some...   

                               additional_references  \
0                                                 {}   
1  {'Encyclopedia Dramatica': 'https://encycloped...   
2  {'Facebook,': 'https://www.facebook.com/search...   
3  {'Wikipedia': 'https://en.wikipedia.org/wiki/C...   
4  {'Meme Generator,': 'https://imgflip.com/meme/...   

                                search_keywords  \
0        [\"This is relevant to my interests\"]   
1                                  [roflcopter]   
2                          [bitches don't know]   
3  [\"leave britney alone\", \"chris crocker\"]   
4                                   [\"o rly\"]   

                                              parent  \
0                                                NaN   
1           https://knowyourmeme.com/memes/lol-ascii   
2                                                NaN   
3  https://knowyourmeme.com/memes/people/chris-cr...   
4  https://knowyourmeme.com/memes/lolspeak-chanspeak   

                                            siblings  \
0                                                NaN   
1  [https://knowyourmeme.com/memes/lollerskates, ...   
2                            

**How many null values?**

In [12]:
memes.isnull().sum()

title                        0
url                          0
last_update_source           0
category                     0
template_image_url           0
meta                         0
ld                           1
added                      157
details                      0
content                      0
tags                         0
additional_references        0
search_keywords           7213
parent                    8232
siblings                  8232
children                 11946
dtype: int64

**Count unique values in cols of Memes**

In [13]:
for col in memes.columns:
    if col in ["meta","ld","details","content","tags","additional_references","search_keywords","siblings","children"]:
        print(col,": nbr of unique values: not countable")
    else:
        print(col,": nbr of unique values: ",len(memes[col].unique()))

title : nbr of unique values:  12654
url : nbr of unique values:  12654
last_update_source : nbr of unique values:  7415
category : nbr of unique values:  1
template_image_url : nbr of unique values:  12654
meta : nbr of unique values: not countable
ld : nbr of unique values: not countable
added : nbr of unique values:  12498
details : nbr of unique values: not countable
content : nbr of unique values: not countable
tags : nbr of unique values: not countable
additional_references : nbr of unique values: not countable
search_keywords : nbr of unique values: not countable
parent : nbr of unique values:  1381
siblings : nbr of unique values: not countable
children : nbr of unique values: not countable


Drop the column category?

**It seems that we have 12654 unique values out of 15406.**

In [14]:
memes.drop_duplicates(subset='title', keep='first')

title  \
0      This is Relevant To My Interests   
1                            ROFLcopter   
2                    Bitches Don't Know   
3                   Leave Britney Alone   
4                                O RLY?   
...                                 ...   
15396           Cee Lo's Muppet Costume   
15397           Liberal Douche Garofalo   
15398                          Polybius   
15400   Donald Trump Reacts to Ted Cruz   
15402            Chocobo Racing Remixes   

                                                     url  last_update_source  \
0      https://knowyourmeme.com/memes/this-is-relevan...          1547002898   
1              https://knowyourmeme.com/memes/roflcopter          1591400337   
2       https://knowyourmeme.com/memes/bitches-dont-know          1607986999   
3      https://knowyourmeme.com/memes/leave-britney-a...          1613129339   
4                   https://knowyourmeme.com/memes/o-rly          1615643899   
...                                                  ...                 ...   
15396  https://knowyourmeme.com/memes/cee-los-muppet-...          1576535826   
15397  https://knowyourmeme.com/memes/liberal-douche-...          1617531761   
15398            https://knowyourmeme.com/memes/polybius          1617051284   
15400  https://knowyourmeme.com/memes/donald-trump-re...          1573134519   
15402  https://knowyourmeme.com/memes/chocobo-racing-...          1555392487   

      category                                 template_image_url  \
0         Meme  https://i.kym-cdn.com/entries/icons/original/0...   
1         Meme  https://i.kym-cdn.com/entries/icons/original/0...   
2         Meme  https://i.kym-cdn.com/entries/icons/original/0...   
3         Meme  https://i.kym-cdn.com/entries/icons/original/0...   
4         Meme  https://i.kym-cdn.com/entries/icons/original/0...   
...        ...                                                ...   
15396     Meme  https://i.kym-cdn.com/entries/icons/original/0...   
15397     Meme  https://i.kym-cdn.com/entries/icons/original/0...   
15398     Meme  https://i.kym-cdn.com/entries/icons/original/0...   
15400     Meme  https://i.kym-cdn.com/entries/icons/original/0...   
15402     Meme  https://i.kym-cdn.com/entries/icons/original/0...   

                                                    meta  \
0      {'og:title': 'This is Relevant To My Interests...   
1      {'og:title': 'ROFLcopter', 'og:site_name': 'Kn...   
2      {'og:title': 'Bitches Don't Know', 'og:site_na...   
3      {'og:title': 'Leave Britney Alone', 'og:site_n...   
4      {'og:title': 'O RLY?', 'og:site_name': 'Know Y...   
...                                                  ...   
15396  {'og:title': 'Cee Lo's Muppet Costume', 'og:si...   
15397  {'og:title': 'Liberal Douche Garofalo', 'og:si...   
15398  {'og:title': 'Polybius', 'og:site_name': 'Know...   
15400  {'og:title': 'Donald Trump Reacts to Ted Cruz'...   
15402  {'og:title': 'Chocobo Racing Remixes', 'og:sit...   

                                                      ld         added  \
0      {'@context': 'http://schema.org', '@type': 'Br...  1.229113e+09   
1      {'@context': 'http://schema.org', '@type': 'Br...  1.229113e+09   
2      {'@context': 'http://schema.org', '@type': 'Br...  1.229113e+09   
3      {'@context': 'http://schema.org', '@type': 'Br...  1.229113e+09   
4      {'@context': 'http://schema.org', '@type': 'Br...  1.229113e+09   
...                                                  ...           ...   
15396  {'@context': 'http://schema.org', '@type': 'Br...  1.297657e+09   
15397  {'@context': 'http://schema.org', '@type': 'Br...  1.332953e+09   
15398  {'@context': 'http://schema.org', '@type': 'Br...  1.314377e+09   
15400  {'@context': 'http://schema.org', '@type': 'Br...  1.469120e+09   
15402  {'@context': 'http://schema.org', '@type': 'Br...  1.334116e+09   

                                                 details  \
0      {'status': 'confirmed', 'origin': 'I Can Has C... 

**So we were right. There were only 12654 rows in memes table. Lets save the result:**

In [15]:
memes=memes.drop_duplicates(subset='title', keep='first')

In [16]:
mainTable = memes.sample(20) # save the samples for excel file

## meta column to meta_data table:

In [17]:
meta_data = pd.DataFrame(memes['meta'].values.tolist())
meta_data.head()

og:title    og:site_name  \
0  This is Relevant To My Interests  Know Your Meme   
1                        ROFLcopter  Know Your Meme   
2                Bitches Don't Know  Know Your Meme   
3               Leave Britney Alone  Know Your Meme   
4                            O RLY?  Know Your Meme   

                                            og:image og:image:width  \
0  https://i.kym-cdn.com/entries/icons/facebook/0...            600   
1  https://i.kym-cdn.com/entries/icons/facebook/0...            600   
2  https://i.kym-cdn.com/entries/icons/facebook/0...            600   
3  https://i.kym-cdn.com/entries/icons/facebook/0...            600   
4  https://i.kym-cdn.com/entries/icons/facebook/0...            600   

  og:image:height  og:type        fb:app_id     fb:pages  \
0             315  article  104675392961482  88519108736   
1             315  article  104675392961482  88519108736   
2             315  article  104675392961482  88519108736   
3             315  article  104675392961482  88519108736   
4             315  article  104675392961482  88519108736   

                       article:publisher         twitter:card   twitter:site  \
0  https://www.facebook.com/knowyourmeme  summary_large_image  @knowyourmeme   
1  https://www.facebook.com/knowyourmeme  summary_large_image  @knowyourmeme   
2  https://www.facebook.com/knowyourmeme  summary_large_image  @knowyourmeme   
3  https://www.facebook.com/knowyourmeme  summary_large_image  @knowyourmeme   
4  https://www.facebook.com/knowyourmeme  summary_large_image  @knowyourmeme   

  twitter:creator                     twitter:title  \
0   @knowyourmeme  This is Relevant To My Interests   
1   @knowyourmeme                        ROFLcopter   
2   @knowyourmeme                Bitches Don't Know   
3   @knowyourmeme               Leave Britney Alone   
4   @knowyourmeme                            O RLY?   

                                 twitter:description  \
0  "This is relevant to my interests" is a popula...   
1  An ASCII helicopter becomes an Internet meme. ...   
2  A braggardly young man flaunts his genitals vi...   
3  “Leave Britney Alone” is a viral video in whic...   
4  O RLY is internet slang for "OH REALLY?" with ...   

                                       twitter:image  \
0  https://i.kym-cdn.com/entries/icons/facebook/0...   
1  https://i.kym-cdn.com/entries/icons/facebook/0...   
2  https://i.kym-cdn.com/entries/icons/facebook/0...   
3  https://i.kym-cdn.com/entries/icons/facebook/0...   
4  https://i.kym-cdn.com/entries/icons/facebook/0...   

                                         description  \
0  "This is relevant to my interests" is a popula...   
1  An ASCII helicopter becomes an Internet meme. ...   
2  A braggardly young man flaunts his genitals vi...   
3  “Leave Britney Alone” is a viral video in whic...   
4  O RLY is internet slang for "OH REALLY?" with ...   

                                              og:url  \
0  https://knowyourmeme.com/memes/this-is-relevan...   
1          https://knowyourmeme.com/memes/roflcopter   
2   https://knowyourmeme.com/memes/bitches-dont-know   
3  https://knowyourmeme.com/memes/leave-britney-a...   
4               https://knowyourmeme.com/memes/o-rly   

                                      og:description  
0  "This is relevant to my interests" is a popula...  
1  An ASCII helicopter becomes an Internet meme. ...  
2  A braggardly young man flaunts his genitals vi...  
3  “Leave Britney Alone” is a viral video in whic...  
4  O RLY is internet slang for "OH REALLY?" with ...

In [18]:
meta_data.shape

(12654, 18)

In [19]:
sum(meta_data.duplicated())

0

In [20]:
for col in meta_data.columns:
    print(col,": nbr of unique values: ",len(meta_data[col].unique()))

og:title : nbr of unique values:  12654
og:site_name : nbr of unique values:  1
og:image : nbr of unique values:  12654
og:image:width : nbr of unique values:  1
og:image:height : nbr of unique values:  1
og:type : nbr of unique values:  1
fb:app_id : nbr of unique values:  1
fb:pages : nbr of unique values:  1
article:publisher : nbr of unique values:  1
twitter:card : nbr of unique values:  1
twitter:site : nbr of unique values:  1
twitter:creator : nbr of unique values:  1
twitter:title : nbr of unique values:  12654
twitter:description : nbr of unique values:  12614
twitter:image : nbr of unique values:  12654
description : nbr of unique values:  12614
og:url : nbr of unique values:  12654
og:description : nbr of unique values:  12614


**We can see that some columns have only 1 unique value for each row. Thus these columns do not give any differentiative information for memes. Let's filter out columns which have different values:** 

In [21]:
metacols = []
for col in meta_data.columns:
    if len(meta_data[col].unique())>1:
        metacols.append(col)

metacols

['og:title',
 'og:image',
 'twitter:title',
 'twitter:description',
 'twitter:image',
 'description',
 'og:url',
 'og:description']

In [22]:
meta_data[metacols].head()

og:title  \
0  This is Relevant To My Interests   
1                        ROFLcopter   
2                Bitches Don't Know   
3               Leave Britney Alone   
4                            O RLY?   

                                            og:image  \
0  https://i.kym-cdn.com/entries/icons/facebook/0...   
1  https://i.kym-cdn.com/entries/icons/facebook/0...   
2  https://i.kym-cdn.com/entries/icons/facebook/0...   
3  https://i.kym-cdn.com/entries/icons/facebook/0...   
4  https://i.kym-cdn.com/entries/icons/facebook/0...   

                      twitter:title  \
0  This is Relevant To My Interests   
1                        ROFLcopter   
2                Bitches Don't Know   
3               Leave Britney Alone   
4                            O RLY?   

                                 twitter:description  \
0  "This is relevant to my interests" is a popula...   
1  An ASCII helicopter becomes an Internet meme. ...   
2  A braggardly young man flaunts his genitals vi...   
3  “Leave Britney Alone” is a viral video in whic...   
4  O RLY is internet slang for "OH REALLY?" with ...   

                                       twitter:image  \
0  https://i.kym-cdn.com/entries/icons/facebook/0...   
1  https://i.kym-cdn.com/entries/icons/facebook/0...   
2  https://i.kym-cdn.com/entries/icons/facebook/0...   
3  https://i.kym-cdn.com/entries/icons/facebook/0...   
4  https://i.kym-cdn.com/entries/icons/facebook/0...   

                                         description  \
0  "This is relevant to my interests" is a popula...   
1  An ASCII helicopter becomes an Internet meme. ...   
2  A braggardly young man flaunts his genitals vi...   
3  “Leave Britney Alone” is a viral video in whic...   
4  O RLY is internet slang for "OH REALLY?" with ...   

                                              og:url  \
0  https://knowyourmeme.com/memes/this-is-relevan...   
1          https://knowyourmeme.com/memes/roflcopter   
2   https://knowyourmeme.com/memes/bitches-dont-know   
3  https://knowyourmeme.com/memes/leave-britney-a...   
4               https://knowyourmeme.com/memes/o-rly   

                                      og:description  
0  "This is relevant to my interests" is a popula...  
1  An ASCII helicopter becomes an Internet meme. ...  
2  A braggardly young man flaunts his genitals vi...  
3  “Leave Britney Alone” is a viral video in whic...  
4  O RLY is internet slang for "OH REALLY?" with ...

**I suspect that there are similar columns in meta_data table:**

In [23]:
sum(meta_data["og:title"]!=meta_data["twitter:title"]) #0
sum(meta_data["og:image"]!=meta_data["twitter:image"]) #0
sum(meta_data["description"]!=meta_data["twitter:description"]) #16
meta_data[metacols][meta_data["description"]!=meta_data["twitter:description"]] # these are just NaN values
sum(meta_data["description"]!=meta_data["og:description"]) #16
meta_data[metacols][meta_data["description"]!=meta_data["og:description"]] # these are just NaN values



og:title  \
1137   Tequila, Heineken, pas ltemps dniaiser   
1895                           Bully Bitch BB   
1970                          Schismatic Pony   
2424                      Gaul Laughing Stock   
5847                        Texting Etiquette   
6793                              Got my nugs   
7240   DJ Jesus - The God of electronic music   
7525                           theflyingcamel   
9762                           Cunt Destroyer   
10311                 Beauty Pageant Reaction   
10877                 Boom Boom Boom Mohammed   
10961                         Van Persie Fall   
11203                        Bad Romance FAIL   
11523                                 My Foot   
11571                             Dancin' Foo   
11744                        Azerbaijan Dance   

                                                og:image  \
1137   https://i.kym-cdn.com/entries/icons/facebook/0...   
1895   https://i.kym-cdn.com/entries/icons/facebook/0...   
1970   https://i.kym-cdn.com/entries/icons/facebook/0...   
2424   https://i.kym-cdn.com/entries/icons/facebook/0...   
5847   https://i.kym-cdn.com/entries/icons/facebook/0...   
6793   https://i.kym-cdn.com/entries/icons/facebook/0...   
7240   https://i.kym-cdn.com/entries/icons/facebook/0...   
7525   https://i.kym-cdn.com/entries/icons/facebook/0...   
9762   https://i.kym-cdn.com/entries/icons/facebook/0...   
10311  https://i.kym-cdn.com/entries/icons/facebook/0...   
10877  https://i.kym-cdn.com/entries/icons/facebook/0...   
10961  https://i.kym-cdn.com/entries/icons/facebook/0...   
11203  https://i.kym-cdn.com/entries/icons/facebook/0...   
11523  https://i.kym-cdn.com/entries/icons/facebook/0...   
11571  https://i.kym-cdn.com/entries/icons/facebook/0...   
11744  https://i.kym-cdn.com/entries/icons/facebook/0...   

                                twitter:title twitter:description  \
1137   Tequila, Heineken, pas ltemps dniaiser                 NaN   
1895                           Bully Bitch BB                 NaN   
1970                          Schismatic Pony                 NaN   
2424                      Gaul Laughing Stock                 NaN   
5847                        Texting Etiquette                 NaN   
6793                              Got my nugs                 NaN   
7240   DJ Jesus - The God of electronic music                 NaN   
7525                           theflyingcamel                 NaN   
9762                           Cunt Destroyer                 NaN   
10311                 Beauty Pageant Reaction                 NaN   
10877                 Boom Boom Boom Mohammed                 NaN   
10961                         Van Persie Fall                 NaN   
11203                        Bad Romance FAIL                 NaN   
11523                                 My Foot                 NaN   
11571                             Dancin' Foo                 NaN   
11744                        Azerbaijan Dance                 NaN   

                                           twitter:image description  \
1137   https://i.kym-cdn.com/entries/icons/facebook/0...         NaN   
1895   https://i.kym-cdn.com/entries/icons/facebook/0...         NaN   
1970   https://i.kym-cdn.com/entries/icons/facebook/0...         NaN   
2424   https://i.kym-cdn.com/entries/icons/facebook/0...         NaN   
5847   https://i.kym-cdn.com/entries/icons/facebook/0...         NaN   
6793   https://i.kym-cdn.com/entries/icons/facebook/0...         NaN   
7240   https://i.kym-cdn.com/entries/icons/facebook/0...         NaN   
7525   https://i.kym-cdn.com/entries/icons/facebook/0...         NaN   
9762   https://i.kym-cdn.com/entries/icons/facebook/0...         NaN   
10311  https://i.kym-cdn.com/entries/icons/facebook/0...         NaN   
10877  https://i.kym-cdn.com/entries/icons/facebook/0...         NaN   
10961  https://i.kym-cdn.com/entries/icons/facebook/0...         NaN   
11203  https://i.kym-cdn.com/entries/icons/facebook/0...         NaN   


**We have only 4 columns with distinctive values in meta_data table:**

In [24]:
meta_data["meta_title"]=meta_data["og:title"]
meta_data["meta_image"]=meta_data["og:image"]
meta_data["meta_desc"]=meta_data["og:description"]
meta_data["meta_url"]=meta_data["og:url"]
newcols = ["meta_title","meta_image","meta_desc","meta_url"]
meta_data[newcols]

meta_title  \
0      This is Relevant To My Interests   
1                            ROFLcopter   
2                    Bitches Don't Know   
3                   Leave Britney Alone   
4                                O RLY?   
...                                 ...   
12649           Cee Lo's Muppet Costume   
12650           Liberal Douche Garofalo   
12651                          Polybius   
12652   Donald Trump Reacts to Ted Cruz   
12653            Chocobo Racing Remixes   

                                              meta_image  \
0      https://i.kym-cdn.com/entries/icons/facebook/0...   
1      https://i.kym-cdn.com/entries/icons/facebook/0...   
2      https://i.kym-cdn.com/entries/icons/facebook/0...   
3      https://i.kym-cdn.com/entries/icons/facebook/0...   
4      https://i.kym-cdn.com/entries/icons/facebook/0...   
...                                                  ...   
12649  https://i.kym-cdn.com/entries/icons/facebook/0...   
12650  https://i.kym-cdn.com/entries/icons/facebook/0...   
12651  https://i.kym-cdn.com/entries/icons/facebook/0...   
12652  https://i.kym-cdn.com/entries/icons/facebook/0...   
12653  https://i.kym-cdn.com/entries/icons/facebook/0...   

                                               meta_desc  \
0      "This is relevant to my interests" is a popula...   
1      An ASCII helicopter becomes an Internet meme. ...   
2      A braggardly young man flaunts his genitals vi...   
3      “Leave Britney Alone” is a viral video in whic...   
4      O RLY is internet slang for "OH REALLY?" with ...   
...                                                  ...   
12649  R&B singer / songwriter Cee Lo is the talk of ...   
12650  Liberal Douche Garofalo is an advice animal im...   
12651  Polybius is a fictional arcade game that was s...   
12652  Ted Cruz refused to endorse Donald Trump at th...   
12653  Chocobo Racing Remixes , also known as “X raci...   

                                                meta_url  
0      https://knowyourmeme.com/memes/this-is-relevan...  
1              https://knowyourmeme.com/memes/roflcopter  
2       https://knowyourmeme.com/memes/bitches-dont-know  
3      https://knowyourmeme.com/memes/leave-britney-a...  
4                   https://knowyourmeme.com/memes/o-rly  
...                                                  ...  
12649  https://knowyourmeme.com/memes/cee-los-muppet-...  
12650  https://knowyourmeme.com/memes/liberal-douche-...  
12651            https://knowyourmeme.com/memes/polybius  
12652  https://knowyourmeme.com/memes/donald-trump-re...  
12653  https://knowyourmeme.com/memes/chocobo-racing-...  

[12654 rows x 4 columns]

In [25]:
md=meta_data[newcols]

**It might be also that** 
- meta_title is same as title
- meta_image is same as template_image_url
- meta_url is same as url
- meta_desc might be something else than other cols in the table...

In [26]:
meta_ex = md.sample(20) #save sample for excel

## ld column to ld_data table, then itemlist, then itemlist_0, itemlist_1

In [27]:
ld_data = pd.json_normalize(memes["ld"])

In [28]:
ld_data["@context"].unique()
ld_data["@type"].unique()
itemlist = pd.json_normalize(ld_data["itemListElement"])

In [29]:
itemlist_0 = pd.json_normalize(itemlist.iloc[:,0])
itemlist_1 = pd.json_normalize(itemlist.iloc[:,1])

In [30]:
itemlist_0["@type"].unique() # 'ListItem', nan
itemlist_0["position"].unique() # 1., nan
itemlist_0["item.@id"].unique() # 'https://knowyourmeme.com/categories/meme', nan
itemlist_0["item.name"].unique() # 'Meme', nan


array(['Meme', nan], dtype=object)

In [31]:
itemlist_1
itemlist_1["@type"].unique() # 'ListItem', nan
itemlist_1["position"].unique() # 2., nan
itemlist_1["item.@id"].unique() # different values - seams like 'url' column
itemlist_1["item.name"].unique() # different values - seams like 'title' column

array(['This is Relevant To My Interests', 'ROFLcopter',
       "Bitches Don't Know", ..., 'Polybius',
       'Donald Trump Reacts to Ted Cruz', 'Chocobo Racing Remixes'],
      dtype=object)

## details column to details table

In [32]:
details = pd.DataFrame(df['details'].values.tolist())
details.head()

status                 origin  year  \
0  confirmed  I Can Has Cheezburger  2006   
1  confirmed           Warcraft III  2004   
2  confirmed                MySpace  2006   
3  confirmed                YouTube  2007   
4  confirmed        Something Awful  2003   

                                           type  
0                                           NaN  
1        [https://knowyourmeme.com/types/slang]  
2  [https://knowyourmeme.com/types/exploitable]  
3                                           NaN  
4                                           NaN

In [33]:
for col in details.columns:
    if col == "type":
        print(col,": nbr of unique values: not countable")
    else:
        print(col,": nbr of unique values: ",len(details[col].unique()))

status : nbr of unique values:  4
origin : nbr of unique values:  6667
year : nbr of unique values:  190
type : nbr of unique values: not countable


In [34]:
details.status.unique()
details.groupby("status").count()

origin   year  type
status                         
confirmed    11888  10940  8069
deadpool      5098   4400   104
submission   11807  10276  6082
unlisted         6      6     1

**Interesting year values:**

In [35]:
year=details.year[details.year.notnull()].astype("int")
year.min() #1000
year.max() #2916
np.sort(year.unique())


array([1000, 1066, 1070, 1100, 1111, 1212, 1232, 1336, 1341, 1392, 1415,
       1445, 1490, 1495, 1503, 1531, 1561, 1564, 1580, 1589, 1590, 1596,
       1600, 1605, 1621, 1700, 1717, 1753, 1774, 1775, 1776, 1780, 1785,
       1790, 1792, 1795, 1797, 1800, 1809, 1814, 1829, 1837, 1839, 1844,
       1846, 1848, 1857, 1861, 1863, 1865, 1867, 1869, 1874, 1876, 1878,
       1880, 1888, 1889, 1890, 1891, 1893, 1894, 1897, 1900, 1902, 1903,
       1904, 1905, 1906, 1907, 1908, 1909, 1910, 1911, 1912, 1913, 1914,
       1915, 1916, 1918, 1919, 1920, 1921, 1922, 1923, 1924, 1925, 1926,
       1927, 1928, 1929, 1930, 1931, 1932, 1933, 1935, 1937, 1938, 1939,
       1940, 1941, 1942, 1943, 1944, 1945, 1946, 1947, 1948, 1949, 1950,
       1951, 1952, 1953, 1954, 1955, 1956, 1957, 1958, 1959, 1960, 1961,
       1962, 1963, 1964, 1965, 1966, 1967, 1968, 1969, 1970, 1971, 1972,
       1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980, 1981, 1982, 1983,
       1984, 1985, 1986, 1987, 1988, 1989, 1990, 19

In [36]:
details.type.sample(20) # I would remove the first part of address: https://knowyourmeme.com/types/

24013                                                  NaN
13932                                                  NaN
8322                                                   NaN
4760              [https://knowyourmeme.com/types/company]
16672                                                  NaN
11841    [https://knowyourmeme.com/types/copypasta, htt...
18664                                                  NaN
3469        [https://knowyourmeme.com/types/advertisement]
653                                                    NaN
17129                                                  NaN
14288         [https://knowyourmeme.com/types/competition]
725                                                    NaN
15477                                                  NaN
4159             [https://knowyourmeme.com/types/emoticon]
9321                                                   NaN
11568                                                  NaN
19383                                                  N

In [37]:
details.type.isnull().sum() # 14543
details.shape # 28799
details["Type"] = details["type"].astype("str")

In [38]:
details.dtypes

status    object
origin    object
year      object
type      object
Type      object
dtype: object

In [39]:
details['Type'] = details['Type'].fillna('[https://knowyourmeme.com/types/]')

In [40]:
details.Type.isnull().sum() # 0

0

In [41]:
details["Type"] = details["Type"].astype("str")
details.dtypes

status    object
origin    object
year      object
type      object
Type      object
dtype: object

In [42]:
Types = details.apply(lambda row: row['Type'].split('/')[-1].strip(), axis=1)

In [43]:
Types

0                  nan
1              slang']
2        exploitable']
3                  nan
4                  nan
             ...      
28794              nan
28795           film']
28796          slang']
28797              nan
28798              nan
Length: 28799, dtype: object

In [44]:
details["Type"] = Types
details.Type
Types = details.apply(lambda row: row['Type'].split("'")[0].strip(), axis=1)

In [45]:
details["Type"] = Types

In [46]:
details["Type"] = details["Type"].astype("str")
details.dtypes

status    object
origin    object
year      object
type      object
Type      object
dtype: object

In [47]:
details.groupby("Type").count()

status  origin  year  type
Type                                     
activist           11      11    10    11
actor             154     154   128   154
advertisement      48      48    47    48
album              12      12    12    12
animal             54      54    44    54
...               ...     ...   ...   ...
visual-effect      37      37    28    37
vlogger           110     110   101   110
web-series         48      48    48    48
webcomic          133     133   133   133
writer             20      20    16    20

[111 rows x 4 columns]

In [48]:
details.Type.isnull().sum() # 0
details[details.Type == 'nan']

status                 origin  year type Type
0       confirmed  I Can Has Cheezburger  2006  NaN  nan
3       confirmed                YouTube  2007  NaN  nan
4       confirmed        Something Awful  2003  NaN  nan
5       confirmed              Happa-Tai  2001  NaN  nan
10      confirmed                Unknown  2002  NaN  nan
...           ...                    ...   ...  ...  ...
28788   confirmed                  4chan  2009  NaN  nan
28791  submission                 Reddit  2016  NaN  nan
28794  submission         Chocobo Racing  2008  NaN  nan
28797   confirmed                Unknown  None  NaN  nan
28798   confirmed                Unknown  None  NaN  nan

[14543 rows x 5 columns]

In [49]:
details.groupby("origin").count()

status  year  type  Type
origin                                                                      
"&" by BMATF                                             1     1     0     1
"99 Problems" by Jay-Z                                   1     1     0     1
"Amending Fences" Episode                                1     1     1     1
"Anger Management"                                       1     1     0     1
"Bad Apple!!" feat. Nomico by Alstroemeria Records       1     1     1     1
...                                                    ...   ...   ...   ...
yukitheater                                              1     1     0     1
zarausdes                                                1     1     0     1
zombies                                                  1     1     0     1
Ōgon Bat                                                 1     1     0     1
‌                                                        1     0     0     1

[6667 rows x 4 columns]

In [50]:
details_examples = details.sample(15)

## content column normalize to norm_content table

In [51]:
norm_content = pd.json_normalize(memes["content"])

In [52]:
norm_content.sample(15)

about.text  \
129    [Diamonds being portrayed as the "hardest meta...   
9201   [Nian Ni (Chinese: 唸你, literally means "Rememb...   
7982                                                 NaN   
7988                                                 NaN   
2102   [Oh, Fuck You Luigi is a misheard quote from S...   
8937                                                 NaN   
1993                                                 NaN   
1155   [Rube Goldberg Machine is a contraption that i...   
3081                                                 NaN   
11922  [Sam was a blind purebred hairless Chinese Cre...   
2127                                                 NaN   
10422                                                NaN   
6250                                                 NaN   
9783                                                 NaN   
4545   [Metal Crotch Guy is the nickname of a perform...   

                                             about.links  \
129    [[GameFAQs, https://knowyourmeme.com/memes/sit...   
9201   [[YouTube, https://knowyourmeme.com/memes/site...   
7982                                                 NaN   
7988                                                 NaN   
2102   [[Super Mario Bros., https://knowyourmeme.com/...   
8937                                                 NaN   
1993                                                 NaN   
1155   [[one take, https://knowyourmeme.com/memes/one...   
3081                                                 NaN   
11922  [[Dog, https://knowyourmeme.com/memes/subcultu...   
2127                                                 NaN   
10422                                                NaN   
6250                                                 NaN   
9783                                                 NaN   
4545   [[viral video, https://knowyourmeme.com/memes/...   

                                             origin.text  \
129    [In 2005 a bunch GameFAQs users had an argumen...   
9201   [The song was composed by the Taiwanese singer...   
7982                                                 NaN   
7988                                                 NaN   
2102   [The episode featuring the quote in question, ...   
8937                                                 NaN   
1993                                                 NaN   
1155   [The original idea was conceived by the Americ...   
3081                                                 NaN   
11922                                                NaN   
2127                                                 NaN   
10422                                                NaN   
6250                                                 NaN   
9783                                                 NaN   
4545   [In 2005, dancer Marie Chouinard[1] began dire...   

                                            origin.links  \
129                                                  NaN   
9201                                                 NaN   
7982                                                 NaN   
7988                                                 NaN   
2102   [[YouTube, https://knowyourmeme.com/memes/site...   
8937                                                 NaN   
1993                                                 NaN   
1155                                                 NaN   
3081                                                 NaN   
11922                                                NaN   
2127                                                 NaN   
10422                                                NaN   
6250                                                 NaN   
9783                                                 NaN   
4545   [[YouTube, https://knowyourmeme.com/memes/site...   

                                             spread.text  \
129    [Some months later, 4chan stumbled upon this a...   
9201   [Initially, the majority of online interest in...   
7982                  

In [53]:
norm_content["notable examples.images"].unique()

TypeError: unhashable type: 'list'

In [ ]:
norm_content.shape

(12654, 15755)

In [ ]:
norm_content.iloc[:,0:9] 
norm_content.iloc[:,2] 
norm_content.iloc[:,4]
norm_content.iloc[:,8]

0        [Search query volume for the keywords "this is...
1        [By September of 2005, searches for the term s...
2        ["Bitches Don't Know" saw a viral takeoff in l...
3                                                      NaN
4        [Search queries for "O rly" peaked in late 200...
                               ...                        
12649                                                  NaN
12650    [Google search queries for "Liberal Douche Gar...
12651                                                  NaN
12652                                                  NaN
12653                                    [[Not Available]]
Name: search interest.text, Length: 12654, dtype: object

In [ ]:
norm_content["About"] = norm_content.iloc[:,0] # 2,4,8
norm_content["Origin"] = norm_content.iloc[:,2] # 2,4,8
norm_content["Spread"] = norm_content.iloc[:,4] # 2,4,8
norm_content["Search_Interest"] = norm_content.iloc[:,8] # 2,4,8


In [ ]:
content = norm_content[["About","Origin","Spread","Search_Interest"]]

## tags

In [ ]:
memes.tags

0        [image macros, comment, comments, approval, im...
1         [ascii, animated gif, rofl, acronym, initialism]
2        [myspace, 4chan, parody, know, bitches, dont, ...
3        [youtube, rant, britney spears, response, frea...
4        [image macro, owl, orly, o rly, negative, some...
                               ...                        
15396    [bird suit, muppets grammys, cee lo, tumblr, v...
15397    [advice dog, hypocrite, liberal, janeane garof...
15398    [game, creepypasta, fanart, nightmare, fangame...
15400    [donald trump, ted cruz, 2016 republican natio...
15402    [ytpmv, final fantasy, chocobo racing, japan, ...
Name: tags, Length: 12654, dtype: object

In [ ]:
Tags = memes[['url','tags']]

In [ ]:
Tags.sample(15)

url  \
145    https://knowyourmeme.com/memes/put-that-cookie...   
11360  https://knowyourmeme.com/memes/gran-turismo-hi...   
12773  https://knowyourmeme.com/memes/majoras-mask-cr...   
13800  https://knowyourmeme.com/memes/unnecessary-cen...   
3312        https://knowyourmeme.com/memes/j-g-wentworth   
4792   https://knowyourmeme.com/memes/doom-2016-cover...   
3240              https://knowyourmeme.com/memes/hurtbae   
3845         https://knowyourmeme.com/memes/xhamsterland   
10524               https://knowyourmeme.com/memes/meter   
4439       https://knowyourmeme.com/memes/dropouthillary   
9047   https://knowyourmeme.com/memes/backseat-lip-du...   
13762        https://knowyourmeme.com/memes/the-sinkhole   
12212       https://knowyourmeme.com/memes/encourage-cat   
3654        https://knowyourmeme.com/memes/montanaspinus   
2503   https://knowyourmeme.com/memes/21-savage-super...   

                                                    tags  
145    [ytmnd, arnold schwarzenegger, music, parody, ...  
11360  [speed, glitches, san, gran, turismo, high, tr...  
12773  [videos, viral, 4chan, /x/, video game, drowne...  
13800  [youtube, edited, beep, bleep, television, cen...  
3312          [cash now, commercial, daytime television]  
4792   [video games, first person shooter, fps, cover...  
3240   [hurt, bae, cheating, relationship, dating, co...  
3845   [fun, humor, video, cartoonn, xhamsterland, st...  
10524                        [meter flood forums bungie]  
4439   [hillary clinton, politics, presidential prima...  
9047    [call me maybe, sports, lip dub, dance, dancing]  
13762                                                 []  
12212                             [encourage cat feline]  
3654             [dinosaurs, saurian, spinosaurus, hoax]  
2503   [rap, 21 savage, espn, twitter, hip hop, music...

In [ ]:
Tags["tags"] = Tags["tags"].astype(str)
Tags2 = Tags.apply(lambda row: row["tags"].split('[')[-1].split(']')[0].split(','), axis=1)
#tags3 = Tags.apply(lambda row: row["tags"].split(']')[0], axis=1)
Tags2.head(20)

/var/folders/hk/kbn1q_sj4jz4lh0m7x05vff80000gn/T/ipykernel_3190/1494570270.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Tags["tags"] = Tags["tags"].astype(str)


0     ['image macros',  'comment',  'comments',  'ap...
1     ['ascii',  'animated gif',  'rofl',  'acronym'...
2     ['myspace',  '4chan',  'parody',  'know',  'bi...
3     ['youtube',  'rant',  'britney spears',  'resp...
4     ['image macro',  'owl',  'orly',  'o rly',  'n...
5                     ['japan',  'yatta',  'happa-tai']
6     ['russia',  'russian reversal',  'image macro'...
7     ['non-english',  'copy & paste',  'photoshop',...
8     ['newscast',  'viral videos',  'catchphrase', ...
9     ['tv',  'commercial',  'mainstream',  'catchph...
10    ['star wars',  'exploitables',  'viral videos'...
11    ['winner',  'winrar',  'engrish',  'pro wrestl...
12    ['remix',  'overdub',  'cartoons',  'viral vid...
13    ['4chan',  'recession raven',  'rich raven',  ...
14    ['animals',  'videos',  'intense',  'staring',...
15    ['catchphrase',  'lotr',  'fantasy',  'pop-cul...
16    ['videos',  'viral',  'popular',  'the interne...
17    ['you have my sword',  'and my bow',  'and

## additional_references

In [ ]:
memes.additional_references.sample(15)

5457     {'Facebook,': 'https://www.facebook.com/TINYBO...
6931                                                    {}
13325                                                   {}
3595                                                    {}
3981     {'Encyclopedia Dramatica,': 'https://encyclope...
10959    {'Urban Dictionary,': 'https://www.urbandictio...
5446                                                    {}
13875                                                   {}
14372    {'Wikipedia': 'https://en.wikipedia.org/wiki/J...
9906     {'Urban Dictionary': 'https://www.urbandiction...
8147     {'Urban Dictionary': 'https://www.urbandiction...
15152    {'Encyclopedia Dramatica,': 'https://encyclope...
14125                                                   {}
13104                                                   {}
9096                                                    {}
Name: additional_references, dtype: object

In [ ]:
References = pd.json_normalize(memes.additional_references)

In [ ]:
References.sample(10)

Encyclopedia Dramatica Facebook,  \
8904                     NaN       NaN   
12298                    NaN       NaN   
7312                     NaN       NaN   
1468                     NaN       NaN   
3188                     NaN       NaN   
10786                    NaN       NaN   
3458                     NaN       NaN   
6979                     NaN       NaN   
8437                     NaN       NaN   
260                      NaN       NaN   

                                        Reddit, Twitter, Urban Dictionary  \
8904   https://www.reddit.com/r/NoShitSherlock/      NaN              NaN   
12298                                       NaN      NaN              NaN   
7312                                        NaN      NaN              NaN   
1468                                        NaN      NaN              NaN   
3188                                        NaN      NaN              NaN   
10786                                       NaN      NaN              NaN   
3458                                        NaN      NaN              NaN   
6979                                        NaN      NaN              NaN   
8437                                        NaN      NaN              NaN   
260                                         NaN      NaN              NaN   

                                               Wikipedia Meme Generator,  \
8904     http://en.wiktionary.org/wiki/no_shit,_Sherlock             NaN   
12298  https://en.wikipedia.org/wiki/User:Fake_Steven...             NaN   
7312                                                 NaN             NaN   
1468                                                 NaN             NaN   
3188                                                 NaN             NaN   
10786                                                NaN             NaN   
3458                                                 NaN             NaN   
6979                                                 NaN             NaN   
8437                 http://pt.wikipedia.org/wiki/Carica             NaN   
260                                                  NaN             NaN   

                                       Urban Dictionary,  \
8904   https://www.urbandictionary.com/define.php?ter...   
12298                                                NaN   
7312                                                 NaN   
1468                                                 NaN   
3188                                                 NaN   
10786                                                NaN   
3458                                                 NaN   
6979                                                 NaN   
8437                                                 NaN   
260                                                  NaN   

      Encyclopedia Dramatica, Wikipedia,  ... Villains Wiki Teh Meme Wiki,  \
8904                      NaN        NaN  ...           NaN            NaN   
12298                     NaN        NaN  ...           NaN            NaN   
7312                      NaN        NaN  ...           NaN            NaN   
1468                      NaN        NaN  ...           NaN            NaN   
3188                      NaN        NaN  ...           NaN            NaN   
10786                     NaN        NaN  ...           NaN            NaN   
3458                      NaN        NaN  ...           NaN            NaN   
6979                      NaN        NaN  ...           NaN            NaN   
8437                      NaN        NaN  ...           NaN            NaN   
260                       NaN        NaN  ...           NaN            NaN   

      WikiFur, Fandom.com Wikisimpsons Smashpedia, dictionary.university  \
8904       NaN        NaN          NaN         NaN                   NaN   
12298      NaN        NaN          NaN         NaN                   NaN   
7312       NaN        NaN          NaN         NaN                   NaN   
1468       NaN        NaN         

In [ ]:
References.count()

Encyclopedia Dramatica    264
Facebook,                 327
Reddit,                   397
Twitter,                  250
Urban Dictionary          768
                         ... 
Smashpedia,                 1
dictionary.university       1
Hoaxes.org,                 1
Snopes,                     1
MOMA,                       1
Length: 144, dtype: int64

## search_keywords

In [ ]:
memes.search_keywords.sample(10)

5772                                      [Doug Dimmadome]
3635     [had it not been for the laws of this land, ha...
360                                                    NaN
7095                         [what character are you quiz]
5741                                            [REEEEEEE]
3778                                                   NaN
12018                                                  NaN
11563                                                  NaN
10849                                                  NaN
12291                                                  NaN
Name: search_keywords, dtype: object

## relatives: parent, siblings, children

In [ ]:
relative_types = ["url","parent","siblings","children"]
memes[relative_types]

url  \
0      https://knowyourmeme.com/memes/this-is-relevan...   
1              https://knowyourmeme.com/memes/roflcopter   
2       https://knowyourmeme.com/memes/bitches-dont-know   
3      https://knowyourmeme.com/memes/leave-britney-a...   
4                   https://knowyourmeme.com/memes/o-rly   
...                                                  ...   
15396  https://knowyourmeme.com/memes/cee-los-muppet-...   
15397  https://knowyourmeme.com/memes/liberal-douche-...   
15398            https://knowyourmeme.com/memes/polybius   
15400  https://knowyourmeme.com/memes/donald-trump-re...   
15402  https://knowyourmeme.com/memes/chocobo-racing-...   

                                                  parent  \
0                                                    NaN   
1               https://knowyourmeme.com/memes/lol-ascii   
2                                                    NaN   
3      https://knowyourmeme.com/memes/people/chris-cr...   
4      https://knowyourmeme.com/memes/lolspeak-chanspeak   
...                                                  ...   
15396                                                NaN   
15397      https://knowyourmeme.com/memes/advice-animals   
15398         https://knowyourmeme.com/memes/creepypasta   
15400  https://knowyourmeme.com/memes/people/donald-t...   
15402  https://knowyourmeme.com/memes/subcultures/fin...   

                                                siblings  \
0                                                    NaN   
1      [https://knowyourmeme.com/memes/lollerskates, ...   
2                                                    NaN   
3      [https://knowyourmeme.com/memes/leave-britney-...   
4      [https://knowyourmeme.com/memes/normie, https:...   
...                                                  ...   
15396                                                NaN   
15397  [https://knowyourmeme.com/memes/angry-advice-m...   
15398  [https://knowyourmeme.com/memes/amandaisherejp...   
15400  [https://knowyourmeme.com/memes/people/mike-li...   
15402  [https://knowyourmeme.com/memes/youre-the-best...   

                                                children  
0                                                    NaN  
1                                                    NaN  
2      [https://knowyourmeme.com/memes/bitches-dont-k...  
3                                                    NaN  
4                                                    NaN  
...                                                  ...  
15396                                                NaN  
15397                                                NaN  
15398                                                NaN  
15400                                                NaN  
15402                                                NaN  

[12654 rows x 4 columns]

In [ ]:
relatives=memes[relative_types]

In [ ]:
s = relatives[["url","siblings"]]
s["siblings"] = s["siblings"].astype("str")
siblings = s.apply(lambda row: row["siblings"].split(','), axis=1).explode()

/var/folders/hk/kbn1q_sj4jz4lh0m7x05vff80000gn/T/ipykernel_3190/470344804.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  s["siblings"] = s["siblings"].astype("str")


In [ ]:
siblings.head(10)

0                                                  nan
1       ['https://knowyourmeme.com/memes/lollerskates'
1         'https://knowyourmeme.com/memes/roflcopter']
2                                                  nan
3    ['https://knowyourmeme.com/memes/leave-britney...
4             ['https://knowyourmeme.com/memes/normie'
4              'https://knowyourmeme.com/memes/oh-hai'
4         'https://knowyourmeme.com/memes/epic-thread'
4      'https://knowyourmeme.com/memes/op-is-a-faggot'
4     'https://knowyourmeme.com/memes/you-win-the-i...
dtype: object

## Creating Excel file

In [ ]:
#with pd.ExcelWriter("examples.xlsx",if_sheet_exists="replace",mode="a") as writer:
#    mainTable.to_excel(writer,sheet_name="mainTable")
#    counts_per_categories.to_excel(writer, sheet_name="CountsPerCategories")
#    meta_ex.to_excel(writer,sheet_name="metaEx")
#    #ld_ex.to_excel(writer,sheet_name="ld_ex")
#    details_examples.to_excel(writer,sheet_name="details_examples")